store/store.go

```diff
@@ -119,6 +125,15 @@ func (s *BoltStore) Init(dataPath string) error {
 	}
 	s.db = db
 
+	// TODO: Having Casbin stuff here doesn't sit quite right with me (especially loading a file here).
+	// Unfortunately we need to re-use the open bolt database for the adapter though.
+	builtinPolicy, _ := assethelper.LoadRBACBuiltinPolicy()
+	casbinAdapter, err := boltadapter.NewAdapter(db, "casbin-policies", builtinPolicy)
```

Comment:
create a casbin store provider. The New should get the db and the policy and the rest of the parameters.
Then in server.go where we initialise the store, you can initialise the casbinProvider. Then pass that provider in where ever we might use / need it. The NewCasbinProvider then can load the file and return an error in case there is a problem which would/should essentially halt starting Gaia since the casbin stuff is part of it even if it is disabled in configs later on.

Makes sense?

Reply:
Hmm. I'm not sure if this would work @Skarlso? We don't have access to the open bolt DB outside of gaia.Store BoltStore. So we wouldn't be able to instantiate a casbin provider inside `server.go` where we depend on the open bolt db.

Reply:
We could move the above code into `service_provider.go` and instantiate the casbin adapter there. Makes more sense. If this error'd the server would fail to launch since we call `store, err := services.StorageService()` in `server.go`.

Reply:
The db is a dependency for the provider. You would create the casbin stuff after we created and returned bolt db.

Reply:
Maybe I'm not following @Skarlso. I don't think the code is setup to do this without some rework. If we're instantiating a `CasbinStoreProvider` within `server.go` we need access to `db *bbolt.DB` there so it can be passed in. That object is privately held within the store logic. We only expose the store interface to `server.go` (rightly so). I don't see a nice way to do this since we create the bolt connection inside `init` rather than passing it in using dependency injection.

Reply:
The casbin store needs to be a separate entity from the gaia store. It does use gaia store though. So implement the two things it needs ( if I see this correctly, there are only 2? ). Extend the Gaia store interface to include policy saving and deleting. Then pass in gaia store as a dependency to casbin store which is using gaia store in the back. For testing casbin we need to be able to inject in a memory based store anyways.

Reply:
No, so the Casbin adapter has multiple methods for read/write (see here - https://github.com/speza/casbin-bolt-adapter/blob/master/adapter.go). It doesn’t use the Gaia store implementation specifically - it only requires the open Bolt connection. I figured we wanted to use the one we already create/open. Once you give it an open bolt connect (and bucket name) the adapter completely takes over.

Let me have a deeper think on this one later mate 😊

Reply:
Oh I see what you mean. Hmmm... Yeah two open connections are not going to work. Bolt doesn't like that. In fact it will throw an error if you try. Alright, we'll think of something. 😊